gateway-service/src/main/java/org/zowe/apiml/gateway/security/service/token/OIDCTokenProvider.java

```diff
@@ -78,72 +69,51 @@ public class OIDCTokenProvider implements OIDCProvider {
     }
 
     @Retryable
-    void fetchJwksUrls() {
+    void fetchJWKSet() {
         if (StringUtils.isBlank(jwksUri)) {
             log.debug("OIDC JWK URI not provided, JWK refresh not performed");
             return;
         }
         log.debug("Refreshing JWK endpoints {}", jwksUri);
-        HttpGet getRequest = new HttpGet(jwksUri);
+
         try {
-            CloseableHttpResponse response = httpClient.execute(getRequest);
-            final int statusCode = response.getStatusLine() != null ? response.getStatusLine().getStatusCode() : 0;
-            final HttpEntity responseEntity = response.getEntity();
-            String responseBody = "";
-            if (responseEntity != null) {
-                responseBody = EntityUtils.toString(responseEntity, StandardCharsets.UTF_8);
-            }
-            if (statusCode == HttpStatus.SC_OK && !responseBody.isEmpty()) {
-                jwks.clear();
-                JwkKeys jwkKeys = mapper.readValue(responseBody, JwkKeys.class);
-                jwks.putAll(processKeys(jwkKeys));
-            } else {
-                log.error("Failed to obtain JWKs from URI {}. Unexpected response: {}, response text: {}", jwksUri, statusCode, responseBody);
-            }
-        } catch (IOException | IllegalStateException e) {
-            log.error("Error processing response from URI {}", jwksUri, e.getMessage());
+            publicKeys.clear();
+            jwkSet = null;
+            jwkSet = JWKSet.load(new URL(jwksUri));
+            publicKeys.putAll(processKeys(jwkSet));
+        } catch (IOException | ParseException | IllegalStateException e) {
+            log.error("Error processing response from URI {} message: {}", jwksUri, e.getMessage());
         }
     }
 
-    private Map<String, Key> processKeys(JwkKeys jwkKeys) {
+
+    private Map<String, Key> processKeys(JWKSet jwkKeys) {
         return jwkKeys.getKeys().stream()
-            .filter(jwkKey -> "sig".equals(jwkKey.getUse()))
-            .filter(jwkKey -> "RSA".equals(jwkKey.getKty()))
-            .collect(Collectors.toMap(JwkKeys.Key::getKid, jwkKey -> {
-                BigInteger modulus = base64ToBigInteger(jwkKey.getN());
-                BigInteger exponent = base64ToBigInteger(jwkKey.getE());
-                RSAPublicKeySpec rsaPublicKeySpec = new RSAPublicKeySpec(modulus, exponent);
+            .filter(jwkKey -> "sig".equals(jwkKey.getKeyUse().getValue()) || "RSA".equals(jwkKey.getKeyType().getValue()))
+            .collect(Collectors.toMap(JWK::getKeyID, jwkKey -> {
                 try {
-                    KeyFactory keyFactory = KeyFactory.getInstance("RSA");
-                    return keyFactory.generatePublic(rsaPublicKeySpec);
-                } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
-                    throw new IllegalStateException("Failed to parse public key");
+                    return jwkKey.toRSAKey().toRSAPublicKey();
```

Comment:
should we check for null pointer exception?

Reply:
toRSAKey() returns "this" so it can't be null